dfs.client.use.datanode.hostname have no effect

The Hadoop cluster equipped with two sets of networks . One is connected to a 10G switch to communicate with servers in the cluster(IP like 192.168.***.***), and the other is connected to a Gigabit switch to connect to client servers outside the cluster(IP like 80.99.***.***).I've tried setting  set up HDFS services (NameNode, DataNode, etc.) to bind to wildcard addresses，and  at the same time setting "dfs.client.use.datanode.hostname" to "true"  in hdfs-client.xml. Before starting clickhouse server，I've setting “LIBHDFS3_CONF=/home/clickhouse/server/conf/hdfs-client.xml”.  I mean I tried passing all the configuration options using those different methods.No effect. And I made sure that the host file resolution configuration of the client server is a Gigabit network IP（such as 80.99.***.***  cdh1.......），but the client server will still access the 10G network through host name resolution，such as 192.168.***.***
![Snipaste_2024-07-10_09-51-16](https://github.com/ClickHouse/libhdfs3/assets/16835515/a2f8b749-8a17-46a8-b7fc-e93fc8a9d41c)
![Snipaste_2024-07-10_09-50-47](https://github.com/ClickHouse/libhdfs3/assets/16835515/84342793-d604-4ccd-b492-fbe6b025b5d2)

> I cannot read the photos.

The content in the picture is the error log, the specific content is:
/user/hive/warehouse/xxx.db/ixxxx_ccrd_label/data_part=1/f149a5589a80e7e0-d78dee46000000931959329948_data.0.parq. Error:HdfsIOException:InputStreamImpl: cannot read file:/user/hive/warehouse/xxx.db/ixxxx_ccrd_label/data part=1/f149a5589a80e7e0-d78dee46000000931959329948_data.0.parq, from position 13153539, size: 1048576. caused by:HdfsI0Exception: InputStreamImpl: all nodes have been tried and no valid replica can be read for Block:[block pool ID:BP-771225177-192.168.xxx.49-1701416672075 block ID. Caused by: HdfsIOException: RemoteBlockReader: Failed to connect to cdhy19y(192.168.xxx.54). Caused by: HdfsTimeoutException: Connect to connect to "192.168.xxx.54:9866" timeout: While executing ParquetBlockinputFormat: While executing HDFSSource.(NETWORK ERROR)(version 22.12.6.22(official build))

--- a/src/client/RemoteBlockReader.cpp
+++ b/src/client/RemoteBlockReader.cpp
@@ -66,7 +66,7 @@ RemoteBlockReader::RemoteBlockReader(const ExtendedBlock& eb,
     readTimeout = conf.getInputReadTimeout();
     writeTimeout = conf.getInputWriteTimeout();
     connTimeout = conf.getInputConnTimeout();
-    sock = getNextPeer(datanode);
+    sock = getNextPeer(datanode, conf.connectViaHostname());
     in = shared_ptr<BufferedSocketReader>(new BufferedSocketReaderImpl(*sock));
     sender = shared_ptr<DataTransferProtocol>(new DataTransferProtocolSender(
         *sock, writeTimeout, datanode.formatAddress()));
@@ -82,14 +82,21 @@ RemoteBlockReader::~RemoteBlockReader() {
     }
 }
 
-shared_ptr<Socket> RemoteBlockReader::getNextPeer(const DatanodeInfo& dn) {
+shared_ptr<Socket> RemoteBlockReader::getNextPeer(const DatanodeInfo& dn, const bool connectToDnViaHostname) {
     shared_ptr<Socket> sock;
     try {
         sock = peerCache.getConnection(dn);
 
         if (!sock) {
             sock = shared_ptr<Socket>(new TcpSocketImpl);
-            sock->connect(dn.getIpAddr().c_str(), dn.getXferPort(),
+            
+            std::string host;
+            if(connectToDnViaHostname) {
+                host = dn.getHostName();
+            } else {
+                host = dn.getIpAddr();
+            }
+            sock->connect(host.c_str(), dn.getXferPort(),
                           connTimeout);
             sock->setNoDelay(true);
         }

--- a/src/client/RemoteBlockReader.h
+++ b/src/client/RemoteBlockReader.h
@@ -76,7 +76,7 @@ public:
 private:
     bool readTrailingEmptyPacket();
     shared_ptr<PacketHeader> readPacketHeader();
-    shared_ptr<Socket> getNextPeer(const DatanodeInfo& dn);
+    shared_ptr<Socket> getNextPeer(const DatanodeInfo& dn, const bool connectToDnViaHostname);
     void checkResponse();
     void readNextPacket();
     void sendStatus();

--- a/src/common/SessionConfig.cpp
+++ b/src/common/SessionConfig.cpp
@@ -69,6 +69,8 @@ SessionConfig::SessionConfig(const Config & conf) {
             &useMappedFile, "input.localread.mappedfile", true
         }, {
             &legacyLocalBlockReader, "dfs.client.use.legacy.blockreader.local", false
+        }, {
+            &connectToDnViaHostname, "dfs.client.use.datanode.hostname", false
         }
     };
     ConfigDefault<int32_t> i32Values[] = {

--- a/src/common/SessionConfig.h
+++ b/src/common/SessionConfig.h
@@ -139,6 +139,10 @@ public:
         return maxLocalBlockInfoCacheSize;
     }
 
+    bool connectViaHostname() const {
+        return connectToDnViaHostname;
+    }
+
     /*
      * OutputStream configure
      */
@@ -339,6 +343,7 @@ public:
     /*
      * InputStream configure
      */
+    bool connectToDnViaHostname;
     bool useMappedFile;
     bool readFromLocal;
     bool notRetryAnotherNode;

--- a/test/data/function-secure.xml
+++ b/test/data/function-secure.xml
@@ -39,5 +39,10 @@
 		<name>dfs.namenode.http-address.gphd-cluster.nn2</name>
 		<value>mdw:50070</value>
 	</property>
+
+	<property>
+		<name>dfs.client.use.datanode.hostname</name>
+		<value>true</value>
+	</property>
 	
 </configuration>

--- a/test/data/function-secure.xml.sample
+++ b/test/data/function-secure.xml.sample
@@ -39,5 +39,10 @@
 		<name>dfs.namenode.http-address.gphd-cluster.nn2</name>
 		<value>mdw:50070</value>
 	</property>
+
+	<property>
+		<name>dfs.client.use.datanode.hostname</name>
+		<value>true</value>
+	</property>
 	
 </configuration>

--- a/test/data/function-test.xml
+++ b/test/data/function-test.xml
@@ -119,4 +119,10 @@
 		<name>dfs.client.read.striped.thread-pool.size</name>
 		<value>64</value>
 	</property>
+
+	<property>
+		<name>dfs.client.use.datanode.hostname</name>
+		<value>true</value>
+	</property>
+	
 </configuration>

--- a/test/data/function-test.xml.sample
+++ b/test/data/function-test.xml.sample
@@ -104,4 +104,10 @@
 		<name>test.get.confint32</name>
 		<value>10</value>
 	</property>
+
+	<property>
+		<name>dfs.client.use.datanode.hostname</name>
+		<value>true</value>
+	</property>
+	
 </configuration>

--- a/test/data/invalidha.xml
+++ b/test/data/invalidha.xml
@@ -26,4 +26,10 @@
                 <name>dfs.client.failover.proxy.provider.phdcluster</name>
                 <value>org.apache.hadoop.hdfs.server.namenode.ha.ConfiguredFailoverProxyProvider</value>
         </property>
+
+	<property>
+		<name>dfs.client.use.datanode.hostname</name>
+		<value>true</value>
+	</property>
+        
 </configuration>

--- a/test/data/validha.xml
+++ b/test/data/validha.xml
@@ -32,4 +32,10 @@
                 <name>dfs.client.failover.proxy.provider.phdcluster</name>
                 <value>org.apache.hadoop.hdfs.server.namenode.ha.ConfiguredFailoverProxyProvider</value>
         </property>
+
+	<property>
+		<name>dfs.client.use.datanode.hostname</name>
+		<value>true</value>
+	</property>
+
 </configuration>

